File: 백준/Silver/9012. 괄호/괄호.cc
#include <iostream>

using namespace std;

int main(){
    int T;
    cin >> T;

    for(int i=0; i<T; i++){
        char stack[51] = {0,};
        int idx = 0;
        string str;
        cin >> str;

        for(int j=0; j<str.length(); j++){
            if(str[j] == '('){
                stack[idx] = '(';
                idx++;
            }
            else{
                if(stack[idx-1] == '('){
                    stack[idx-1] = 0;
                    idx--;
                }
                else{
                    stack[idx] = ')';
                    idx++;
                }
            }
        }

        if(stack[0] == 0)
            cout << "YES" << endl;
        else
            cout << "NO" << endl;
    }

    return 0;
}

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine());
        Stack<Character> st = new Stack<>();
        Boolean isVps = true;

        for(int i=0; i<T; i++){
            isVps = true;
            st = new Stack<>();
            String str = br.readLine();

            for(int j=0; j<str.length(); j++){
                // JAVA는 str[i] 대신 str.charAt(i) 사용
                if(str.charAt(j) == '(')
                    st.push('(');
                else{
                    if(st.isEmpty()){
                        isVps = false;
                        break;
                    }
                    else st.pop();
                }
            }

            if(st.isEmpty() && isVps)
                System.out.println("YES");
            else
                System.out.println("NO");
        }
    }
}
